### PR TITLE
#202: Additive expression does not handle union{int,number} correctly

### PR DESCRIPTION
**Pull request by jpilgrim**
#202

- additive expression with "add" now takes composed types, boolean and null into account to infer return type to number, string, or union{number,string}
- additive, multiplication and shift expressions now accept any type -- further checks need to be done by non-type system analysis
- improved tests (use "xpect type of" now instead of "xpect errors")

### CODE REVIEW

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/messages.properties [hunk 1]
```diff
@@ -624,8 +624,8 @@ EXP_CAST_FAILED=error;;;Cannot cast from {0} to {1}
 EXP_CAST_UNNECESSARY=warning;;;Unnecessary cast from {0} to {1}
 # no parameter required
 EXP_CAST_INVALID_TARGET=error;;;Can only cast to class, interface, enum, function, primitive, union or intersection types
-# 0: recognized type
-EXP_IS_NOT_A_VALID_NUMBER=error;;;{0} is not a valid number.
+# 0: recognized type, 1: expression type
+EXP_RESULT_IS_NAN=warning;;;Operand of type {0} in {1} will always result in NaN.
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/messages.properties]: The actual reason for getting `NaN` is, in most cases, not the type of the operand but the actual value of the operand. Even if those "most cases" are not implemented yet, I'd rephrase the message to:
```
An operand of {0} in {1} will always result in NaN.
```

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSExpressionValidator.xtend [hunk 2]
```diff
@@ -874,38 +876,51 @@ class N4JSExpressionValidator extends AbstractN4JSDeclarativeValidator {
...
-				return; // corrupt AST (e.g., while editing)
-			}
-
-			if (trhs.declaredType === bits.nullType || trhs.declaredType === bits.undefinedType)
-				issueNotANumberType(trhs.declaredType.name, ae.rhs);
+		val tlhs = ts.tau(lhs)
+		if (tlhs === null) {
+			return; // corrupt AST (e.g., while editing)
+		}
+		if (tlhs.declaredType === bits.undefinedType)
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSExpressionValidator.xtend]: why not already support value `null` as well (i.e. show warning if declaredType if `null` type)?

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/n4js.xsemantics [hunk 3]
```diff
@@ -958,18 +958,25 @@ from {
...
 			} else {
-				if ( (lunknown || runknown) && (lnum || rnum || undef)  ) {
-					T = if (lunknown) l else r
+				if ( (lunknown || runknown) && (lnum || rnum)  ) {
+					// one is numeric, the other unknown --> 
+					T = typeSystemHelper.createUnionType(G, G.numberTypeRef, G.stringTypeRef);
 				} else {
-					T = G.stringTypeRef
+					val lMayNum = lnum || G.containsNumericOperand(l);
+					val rMayNum = rnum || G.containsNumericOperand(r);
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/n4js.xsemantics]: `lMayNum`/`rMayNum` must be true also if `l`/`r` is a super type of `number` (i.e. `any`).

At the moment, the following code breaks:
```
let a: any = 5;
let str: string = 40 + a;  // does not show error
str.trim();  // exception at runtime
```

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/n4js.xsemantics [hunk 4]
```diff
@@ -2007,33 +2014,21 @@ from {
...
 /* @see [N4JS] 6.1.14 Additive Expression */
 rule expectedTypeInAdditiveExpression
 	G |- AdditiveExpression e |> Expression expression : TypeRef T
 from {
-	if (e.op != AdditiveOperator.ADD && jsVariantHelper.isTypeAware(e)) {
-		T = G.numberTypeRef
-	} else {
-		T = G.anyTypeRef
-	}
+	T = G.anyTypeRef
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/n4js.xsemantics]: type `symbol` is disallowed, e.g. the following throws an exception at runtime:
```
let sym: symbol = Symbol.iterator;
sym+5;
```

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/messages.properties (hunk 1 above)
Comment (jpilgrim) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/messages.properties]: Updated the message template, but the message will remain the same since the validation stems from a type check.

We will probably replace these checks with data-analysis related checks anyway.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSExpressionValidator.xtend (hunk 2 above)
Comment (jpilgrim) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSExpressionValidator.xtend]: I added a waring that null is interpreted as 0. Note that

```
1/null != 0/null
```

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/n4js.xsemantics (hunk 3 above)
Comment (jpilgrim) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/n4js.xsemantics]: Good catch. Added

```
// XPECT type of a+n --> union{number,string}
a+n
// XPECT type of a+s --> string
a+s
```

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/n4js.xsemantics (hunk 4 above)
Comment (jpilgrim) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/n4js.xsemantics]: I added

```
let sym: symbol = Symbol.iterator;
// XPECT errors --> "Operand of type symbol cannot be converted to number." at "sym"
// XPECT type of n+sym --> union{number,string}
n+sym
// XPECT errors --> "Operand of type symbol cannot be converted to number." at "sym"
// XPECT type of sym+n --> union{number,string}
sym+n
// XPECT errors --> "Operand of type symbol cannot be converted to number." at "sym"
// XPECT type of s+sym --> string
s+sym
// XPECT errors --> "Operand of type symbol cannot be converted to number." at "sym"
// XPECT type of sym+s --> string
sym+s
```
(and similar for other operations, in these cases always expecting number).

I also improved the type inference of "+": I will never return "unknown type" but union{number,string} instead.

**Comment by jpilgrim** [comment]
@mor-n4 Fixed issues, please have a quick look.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/n4js.xsemantics (hunk 4 above)
Comment (mor-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/n4js.xsemantics]: > I also improved the type inference of "+": I will never return "unknown type" but union{number,string} instead.

Not sure about this one, because iirc we had the rule if an inference that depends on an unknown type should result to "unknown type" (i.e. don't suppress "unknown type" but pass on though hierarchy). But not sure right now. Would also be ok to keep as is.

**Comment by mor-n4** [comment]
aside from my minor remark about "unknown type", all good!

:+1:
